witness/src/Dscp/Witness/Web/Logic.hs

```diff
@@ -246,9 +247,18 @@ checkFairCVPDF
     => Pdf.PDFBody -> m FairCVAndCheckResult
 checkFairCVPDF pdf = do
     let maybeFairCV = do
-            fairCVencoded <- Pdf.project (Pdf.MaxSearchLength Nothing) pdf
-            Aeson.decodeStrict fairCVencoded
+            (fairCVencoded, Pdf.PDFBody source) <- Pdf.unInject (Pdf.MaxSearchLength Nothing) pdf
+            res <- Aeson.decodeStrict fairCVencoded
+            return (res, source)
+
+    (fairCV, source) <- maybe (throwM InvalidFormat) pure maybeFairCV
+    checkRes         <- checkFairCV fairCV
+
+    let pdfHashIsValid =
+            (  fairCV^?to(fcCV).each.each.to(toList).each.ptSignedSubmission.ssSubmission.sContentsHash
+            == Just (coerce $ hash source)
+            )
 
-    fairCV   <- maybe (throwM InvalidFormat) pure maybeFairCV
-    checkRes <- checkFairCV fairCV
     return $ FairCVAndCheckResult fairCV checkRes
+        { fairCVFullyValid = fairCVFullyValid checkRes && pdfHashIsValid
```

Comment:
I believe this is a pretty confusing way to report verification error :cat: (however I dunno what to suggest here, throwing a custom exception seems at least to be not worse).
Calling @flyingleafe for comment.